docs(importing): docker-compose import (#380)
btpworks/chronicle: docs/importing.md

# Importing existing Data into Chronicle

## Importing Data into Chronicle

Chronicle provides a CLI command `import` to load data from a JSON-LD file that
contains an array of Chronicle Operations. These operations can be used to
create agents, namespaces, and other resources in the Chronicle database. For
more information about the data format of the import file, see [Recording Provenance](../recording_provenance/#importing-data-into-chronicle).

### Local development environment

To import data into Chronicle, follow these steps:

- Create a JSON-LD file with an array of Chronicle Operations.
- Save the file in the Chronicle root directory with a unique name, for example,
  `import.json`.
- Run the following command:

  ```bash
  chronicle --bin chronicle -- import <namespace-id> <namespace-uuid> <json-file-url>
  ```

> If the data is successfully imported, the Chronicle Operations will be added

Reviewer: Could we maybe DRY this up a bit and remove and/or reference the documentation getting duplicated here in `cli.md`, for example, as well as in `recording_provenance.md`?

Author: I've removed the references here, and linked to the CLI docs

to the Chronicle database.

### Docker compose environment

To import data into Chronicle, follow these steps:

- Create a JSON-LD file with an array of Chronicle Operations.
- Add a bind mount to your docker compose file mounting your import file in the
  chronicle container, if you're using the chronicle-examples or
  chronicle-bootstrap configuration then this will for the

>   `chronicle-sawtooth-api` container, defined in  `docker/chronicle.yaml`.

Reviewer 2: If they're running `make run-my-domain` without `stl` in there then it'll be `docker/chronicle-domain.yaml` that they need to edit instead.

  ```yaml
  volumes:
    - type: bind
      source: ~/import.json
      target: /import.json
  ```

- Start your docker compose environment
- Exec into the chronicle container

  ```bash

>   docker exec -it chronicle bash

Reviewer 2: Might need to make it clearer that this probably won't work as-is, e.g., I needed to substitute the name for `docker_chronicle-sawtooth-api_1` in testing.

Author: I've added a note about this

## Example import process

### Local development environment

To import data into Chronicle, follow these steps:

- Create a JSON-LD file with an array of Chronicle Operations.
- Save the file in the Chronicle root directory with a unique name, for example,
  `import.json`.
- Run the following command:

  ```bash
  chronicle --bin chronicle -- import <namespace-id> <namespace-uuid> <json-file-url>
  ```

See the [CLI `import` command help](../recording_provenance/#importing-data-into-chronicle)
for more information on each argument in the command.

If the data is successfully imported, the Chronicle Operations will be added
to the Chronicle database.

### Docker compose environment

Note: This example assumes you are using the chronicle-examples and running the
stl version of the docker compose file. You will need to adapt the instructions
to your environment, specifically the name of the chronicle container.

To import data into Chronicle, follow these steps:

- Create a JSON-LD file with an array of Chronicle Operations.
- Add a bind mount to your docker compose file mounting your file in the

>   chronicle container. This will be f you're using the chronicle-examples or
>   chronicle-bootstrap configuration then this will be for the
>   `chronicle-sawtooth-api` container, defined in  `docker/chronicle.yaml`.

Reviewer: ```suggestion
  chronicle container. If you're using [chronicle-examples](https://github.com/btpworks/chronicle-examples/)
  or [chronicle-bootstrap](https://github.com/btpworks/chronicle-bootstrap) then this will be the
  `chronicle-sawtooth-api` container, defined in  `docker/chronicle.yaml`.
```

# Importing existing Data into Chronicle

## Importing when running on Docker compose

## Importing Data into Chronicle

Chronicle provides a CLI command `import` to load data from a JSON-LD file that
contains an array of Chronicle Operations. These operations can be used to
create agents, namespaces, and other resources in the Chronicle database.

### Local development environment

To import data into Chronicle, follow these steps:

- Create a JSON-LD file with an array of Chronicle Operations.
- Save the file in the Chronicle root directory with a unique name, for example,
  `import.json`.
- Run the following command:

  ```bash
  chronicle --bin chronicle -- import <namespace-id> <namespace-uuid> <json-file-url>
  ```

  Where:

>   `<namespace-id>`: the name of the namespace to which the data will be imported.

Reviewer 2: This lines all run together when rendered.

  `<namespace-uuid>`: the UUID of the namespace to which the data will be imported.
  `<json-file-url>`: the URL of the JSON-LD file containing the Chronicle Operations.

- If the data is successfully imported, the Chronicle Operations will be added
  to the Chronicle database.

### Docker compose environment

To import data into Chronicle, follow these steps:

- Create a JSON-LD file with an array of Chronicle Operations.
- Add a bind mount to your docker compose file mounting your import file in the

>   chronicle container, if you're using the chronicle-examples or chronicle
>   bootstrap configuration this will in the `docker/chronicle.yaml`
>   and is the `chronicle-sawtooth-api` container.

Reviewer 2: ```suggestion
  chronicle container, if you're using the chronicle-examples or
  chronicle-bootstrap configuration then this will be for the
  `chronicle-sawtooth-api` container, defined in `docker/chronicle.yaml`.
```

Author: Thanks, I've updated the text

  or [chronicle-bootstrap](https://github.com/btpworks/chronicle-bootstrap)
  configuration then this will be for the `chronicle-sawtooth-api` container,
  as defined in  `docker/chronicle.yaml`.

  ```yaml
  volumes:
    - type: bind

>       source: ~/import.json

Reviewer: ```suggestion
      source: ./import.json
```

I was just working on adding support for `import` to examples and caught that earlier you say, "Save the file in the Chronicle root directory," but here it's saved to `~/`. 

It might be better to update `chronicle-examples` with an empty `import.json` file in the root of the `chronicle-examples` repo like I've done in [this branch](https://github.com/btpworks/chronicle-examples/compare/main...feature/make-import-work-with-examples). That way we can just add the bind mount to the source code.

Reviewer 2: I don't think we need the tracked blank file, the instructions have us place a file before starting docker. Still, I'd agree that the instructions *do* need a bit of fixing here to ensure that the `import.json` is placed where the compose file expects, well-spotted.

Author: Makes sense to clarify it more, I've updated the text and added note about changing it for your own environment. What do you think?

- Start your docker compose environment
- Exec into the chronicle container

  ```bash
  docker exec -it chronicle bash
  ```

- Run the following command:

  ```bash
    chronicle -c /etc/chronicle/config/config.toml \
    --console-logging pretty \
    --sawtooth tcp://validator:4004 \
    import <namespace-id> <namespace-uuid> file:///import.json

>     ```

Reviewer 2: The rendering around here doesn't look to be going to plan.

>     ```

Reviewer 2: The rendering around here doesn't look to be going to plan.

    --sawtooth tcp://validator:4004 \
    import <namespace-id> <namespace-uuid> file:///import.json
  ```

> See the  [CLI `import` command help](../recording_provenance/#importing-data-into-chronicle)

Reviewer 2: Was this link meant to be into `cli.md`?

Author: Thanks yes it should have been, I've updated the link

Author: @mtbc happy for me to resolve this?

Reviewer 2: Missed in in looking through again, thanks for asking.

# Importing existing Data into Chronicle

## Importing Data into Chronicle

Chronicle provides a CLI command `import` to load data from a JSON-LD file that
contains an array of Chronicle Operations. These operations can be used to
create agents, namespaces, and other resources in the Chronicle database. For

> more information about the data format of the import file, see [Recording Provenance](../cli#import-namespace-id-namespace-uuid-url).

Reviewer: ```suggestion
more information about the data format of the import file, see [Recording Provenance](recording_provenance.md#importing-data-into-chronicle).
```
Sorry, @alexhq, found another thing: The info about the data format etc is in Recording Provenance in the Importing Data into Chronicle section.

Author: Ah thanks, I've corrected that now

  `import.json`.
- Run the following command:

  ```bash
  chronicle --bin chronicle -- import <namespace-id> <namespace-uuid> <json-file-url>
  ```

See the [CLI `import` command help](../cli)
for more information on each argument in the command.

If the data is successfully imported, the Chronicle Operations will be added
to the Chronicle database.

### Docker compose environment

Note: This example assumes you are using the chronicle-examples and running the
stl version of the docker compose file. You will need to adapt the instructions
to your environment, specifically the name of the chronicle container.

To import data into Chronicle, follow these steps:

- Create a JSON-LD file with an array of Chronicle Operations.
- Add a bind mount to your docker compose file mounting your file in the
  chronicle container. If you're using the [chronicle-examples](https://github.com/btpworks/chronicle-examples/)
  or [chronicle-bootstrap](https://github.com/btpworks/chronicle-bootstrap)
  configuration then this will be for the `chronicle-sawtooth-api` container,
  as defined in  `docker/chronicle.yaml`.

Note: You will need to adapt the instructions to your environment, specifically

> the path to your import source file. The following example assumes that the you

Reviewer: ```suggestion
the path to your import source file. The following example assumes that you
```

are using the [chronicle-examples](https://github.com/btpworks/chronicle-examples/)
repository and that your import file is in the root directory of the repository.

  ```yaml
  volumes:
    - type: bind
      source: ../import.json
      target: /import.json
  ```

- Start your docker compose environment
- Exec into the chronicle container

  ```bash
  docker exec -it chronicle bash
  ```

- Run the following command:

  ```bash
    chronicle -c /etc/chronicle/config/config.toml \
    --console-logging pretty \
    --sawtooth tcp://validator:4004 \
    import <namespace-id> <namespace-uuid> file:///import.json
  ```

> See the  [CLI `import` command help](../cli)

Reviewer: ```suggestion
See the  [CLI `import` command help](../cli#import-namespace-id-namespace-uuid-url)
```

Author: Thanks, I've updated all the cli links with the anchor
